### ISSUE TITLE
Console dimensions are never set using createTerminal(options: ExtensionTerminalOptions)

### ISSUE THREAD

**Issue by GabeDeBacker** [microsoft/vscode#107239]
Version: 1.49.1 (user setup)
Commit: 58bb7b2331731bf72587010e943852e13e6fd3cf
Date: 2020-09-16T23:27:51.792Z
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Windows_NT x64 10.0.20215

Steps to Reproduce:
Have an extension create a terminal.
```typescript
        this.terminal = vscode.window.createTerminal({
            name: this.terminalName, 
            pty: { 
                    onDidWrite: this.razzleOutputStringEmitter.event,
                    close: this.onTerminalClosed.bind(this),
                    open: this.onTerminalOpened.bind(this),
                    setDimensions: this.onSetTerminalDimensions.bind(this)
            }
        });```

    private onTerminalOpened(initialDimensions: vscode.TerminalDimensions | undefined) : void {
        this.terminaldDimensions = initialDimensions || DefaultPseudoConsoleDimensions;
    }

    private onSetTerminalDimensions(dimensions: vscode.TerminalDimensions): void {
        this.terminaldDimensions = dimensions;
```

When the terminal is shown (either programmatically or via UI) the terminal dimensions on both callbacks come in as undefined.

Due to this, we cannot set the proper dimensions on the Windows CON-PTY conhost.exe and that results in corrupted output. 

Does this issue occur when all extensions are disabled?: Yes

**Comment by GabeDeBacker** [comment]
@vscode-triage-bot  - No, this is not a duplicate

**Comment by meganrogge** [comment]
to verify:

- clone https://github.com/microsoft/vscode-extension-samples/
- create a symlink to vscode-extension-samples/extension-terminal-sample within your .vscode-oss-dev directory. 
- open a vscode window -> open folder -> open the symlink pointing to vscode-extension-samples/extension-terminal-sample
- run npm install
- add the following code to extension.ts within the vscode-extension-samples/extension-terminal-sample folder:

```
	context.subscriptions.push(vscode.commands.registerCommand('extensionTerminalSample.createAndResize', async () => {
		const dimensionsEmitter = new vscode.EventEmitter<any>();
		const pty = {
			onDidWrite: writeEmitter.event,
			onDidOverrideDimensions: dimensionsEmitter.event,
			open: () => {
				dimensionsEmitter.fire({
					columns: 20,
					rows: 10
				});
			},
			setDimensions: d => console.log('setDimensions updated', d)
		};
		const terminal = (<any>vscode.window).createTerminal({ name: 'My terminal', pty });
		terminal.creationOptions.cwd;
		terminal.show();
	}));

	function delay(ms: number) {
		return new Promise(resolve => setTimeout(resolve, ms));
	}

	context.subscriptions.push(vscode.commands.registerCommand('extensionTerminalSample.clear', () => {
		writeEmitter.fire('\x1b[2J\x1b[3J\x1b[;H');
	}));
}
```
- build and launch VS Code
- in OSS, open Dev tools to reveal the console
- open the command palette in OSS.
- if your symlink worked, you will see the create and resize command in the command pallette. run this.
- in the console of chrome dev tools, confirm that setDimensions is logged with the dimensions immediately when the terminal is created.

**Comment by JacksonKearl** [comment]
Not sure this is fixed. I have extension code (added to the terminal sample extension):
```ts
	context.subscriptions.push(vscode.commands.registerCommand('extensionTerminalSample.create', async () => {
		const dimensionsEmitter = new vscode.EventEmitter<any>();
		const writeEmitter = new vscode.EventEmitter<string>();

		const pty: vscode.Pseudoterminal = {
			onDidWrite: writeEmitter.event,
			onDidOverrideDimensions: dimensionsEmitter.event,
			close: () => { },
			open: (dimensions) => {
				console.log('open', dimensions)
				dimensionsEmitter.fire({
					columns: 20,
					rows: 10
				});
			},
			setDimensions: d => console.log('setDimensions updated', d)
		};
		const terminal = vscode.window.createTerminal({ name: 'My terminal', pty });
		terminal.show();
	}));
```

When I run this... 

In Stable I get
```
[Extension Host] Terminal opened. Total count: 1 (at /Users/jkearl/Developer/vscode-extension-samples/terminal-sample/out/extension.js:29:17)
console.ts:137 [Extension Host] open undefined (at Object.open (/Users/jkearl/Developer/vscode-extension-samples/terminal-sample/out/extension.js:16:25))
console.ts:137 [Extension Host] Active terminal changed, name=My terminal (at /Users/jkearl/Developer/vscode-extension-samples/terminal-sample/out/extension.js:36:17)
console.ts:137 [Extension Host] setDimensions updated {columns: 50, rows: 22} (at Object.setDimensions (/Users/jkearl/Developer/vscode-extension-samples/terminal-sample/out/extension.js:22:41))
```

And in Insiders I get 
```
[Extension Host] Terminal opened. Total count: 1 (at /Users/jkearl/Developer/vscode-extension-samples/terminal-sample/src/extension.ts:33:11)
workbench.desktop.main.js:60 [Extension Host] Active terminal changed, name=My terminal (at /Users/jkearl/Developer/vscode-extension-samples/terminal-sample/src/extension.ts:41:11)
workbench.desktop.main.js:60 [Extension Host] open undefined (at Object.open (/Users/jkearl/Developer/vscode-extension-samples/terminal-sample/src/extension.ts:19:13))
```

So in Insiders the `setDimensions` hook isn't called on startup, and in both the `open` hook doesn't get passed any dimensions. I think the open hook not receiving dimensions may be intentional, not sure. But I believe `setDimensions` ought to be called at startup?

Note: I just verified this by building the extension as normal, is the whole symlink setup really required? Standard extensions should see the correct behaviour too yes?

**Comment by Tyriar** [comment]
The symlink setup was needed for running out of sources, regular ext debug should work when the fix is in insiders.

**Comment by JacksonKearl** [comment]
Okay that's what I suspected. Verification-found still in effect.

**Comment by meganrogge** [comment]
@JacksonKearl are you not seeing this?

![recording (13)](https://user-images.githubusercontent.com/29464607/106309599-35b8b780-6217-11eb-9517-d171e2ff95f9.gif)

**Comment by JacksonKearl** [comment]
Hm, it's sporadic. About half the time I don't get the initial `setDimensions` call. See here:
![recording (13)](https://user-images.githubusercontent.com/8586769/106311603-4159ad80-621a-11eb-82f7-81d286e165ac.gif)

**Comment by meganrogge** [comment]
Yes it seems to only work on terminal creation when it's:
- the first terminal tab
- a split tab

**Comment by meganrogge** [comment]
It looks like there's a regression related to the terminal reconnect work related to [these lines
](https://github.com/microsoft/vscode/blob/a095d7fcf74d0b072886c3041bd5932f3b5be445/src/vs/workbench/contrib/terminal/browser/terminalInstance.ts#L1481-L1485)

**Comment by meganrogge** [comment]
![recording - 2021-05-14T134135 410](https://user-images.githubusercontent.com/29464607/118326670-20138900-b4ba-11eb-9396-c5e3002b27fa.gif)
This now works for all terminals that get created.

**Comment by meganrogge** [comment]
Closing as fixed by 1d1b902c11a661184326f4ec8dd123e2e5726aea
